### PR TITLE
implement boxv2

### PR DESCRIPTION
**Pull request by mlsteele**
Implement `boxV2` and `unboxV2`. Writing with `boxV2` is not used outside of tests.

This is the first client that can read V2 messages. So, if all goes well, once this version of the client has been out for long enough (~1 month) then we can enable writing V2 messages.

Sprinkled `CORE-4540` notes around for when in about a month we enable boxV2.

This has a `compareHeaders` method that is very similar to [this](https://github.com/keybase/client/pull/6044/files#diff-d1325d9a259c7d25092218f3fa0661d4R832). So I'll deal with that after that PR lands.

The plan:
- Release this client which can read v2.
- Wait a month.
- Critical out clients that can't read v2.
- Release a client in which
  - Sends only v2 messages
  - `HeaderPlaintextV1` has a *optional* `MerkleRoot` field added to it. Once we're not sending any more `MessageBoxedV1` messages, it should be safe to add `omitempty` fields to `HeaderPlaintextV1` without breaking V1 unboxing. V1 unboxing is finicky because it reserializes before checking the sig.
  - Set that MerkleRoot

I'll test that out manually before merging this.

### CODE REVIEW

File: go/chat/boxer_test.go [hunk 1]
```diff
@@ -845,13 +858,11 @@ func TestV1Message3(t *testing.T) {
...
 	require.Equal(t, false, unboxed.ClientHeader.TlfPublic)
 	require.Equal(t, chat1.MessageType_TEXT, unboxed.ClientHeader.MessageType)
-	require.Equal(t, chat1.MessageID(0), unboxed.ClientHeader.Supersedes)
-	require.Nil(t, unboxed.ClientHeader.Deletes)
 	expectedPrevs := []chat1.MessagePreviousPointer{chat1.MessagePreviousPointer{Id: 2, Hash: chat1.Hash{0xbe, 0xb2, 0x7c, 0x41, 0xdb, 0xeb, 0x2e, 0x90, 0x04, 0xf2, 0x48, 0xf2, 0x78, 0x24, 0x3a, 0xde, 0x5e, 0x12, 0x0c, 0xb7, 0xc4, 0x1f, 0x40, 0xe8, 0x47, 0xa2, 0xe2, 0x2f, 0xe8, 0x2c, 0xd3, 0xb4}}}
 	require.Equal(t, expectedPrevs, unboxed.ClientHeader.Prev)
 	require.Equal(t, canned.SenderUID(t), unboxed.ClientHeader.Sender)
 	require.Equal(t, canned.SenderDeviceID(t), unboxed.ClientHeader.SenderDevice)
-	require.Nil(t, unboxed.ClientHeader.MerkleRoot)
+	// require.Nil(t, unboxed.ClientHeader.MerkleRoot)
```
Comment (oconnor663) [inline go/chat/boxer_test.go]: Accidental commit? Or maybe needs a comment?

Comment (mlsteele) [inline go/chat/boxer_test.go]: Needs a comment. This field isn't in `HeaderVerified` anymore but will be after `CORE-4540`. I'll add those comments in

**Comment by mlsteele** [comment]
Note to self: MBv2 needs a gregor-side change. Because the `MessageBoxed` is split up on the gregor side, it needs to not throw out the new `headerSealed`. The tests for this PR appear to have passed despite this, which is not great, so I'll look into that too.

File: go/chat/boxer.go [hunk 2]
```diff
@@ -827,42 +1223,69 @@ func (b *Boxer) unmarshal(data []byte, v interface{}) error {
...
+
+	// SenderDevice
+	if !bytes.Equal(hServer.SenderDevice.Bytes(), hSigned.SenderDevice.Bytes()) {
+		return NewPermanentUnboxingError(NewHeaderMismatchError("SenderDevice"))
 	}
-	// TODO: Supersedes and other fields currently don't match after
-	// decryption. The easiest thing might be to stop using the same data
-	// structure on both sides.
+
+	// CORE-4540: _Don't_ enable this check! There are V1 messages in the wild with
```
Comment (oconnor663) [inline go/chat/boxer.go]: "this check" = which check?

File: go/chat/boxer.go [hunk 3]
```diff
@@ -402,8 +412,276 @@ func (b *Boxer) unboxV1(ctx context.Context, boxed chat1.MessageBoxed, encryptio
...
+
+// Compare the unsigned and signed header for MessageBoxedVersion_V1.
+// The V1 and V2 checks are different methods because they are strict on slightly different things.
+// Confirm that fields in the server-supplied ClientHeader match what
+// we decrypt. It would be preferable if the server didn't supply this data
+// at all (so that we didn't have to worry about anyone trusting it
+// *before* we get to this check, for example), but since we have it we
+// need to check it.
+// The most important check here is that the Sender and SenderDevice match.
+// That is the only thing that gives the verification key used credibility.
```
Comment (oconnor663) [inline go/chat/boxer.go]: Confirming my understanding of the tricky details here: It would _almost_ be ok for us to ignore the Sender entirely here, and instead look everything up via the signing KID. However, that opens us up to a weird attack where someone delegates a crazy weak sibkey that can return true for any signature, and then with evil Keybase's cooperation they claim authorship of other people's messages in an existing convo (that they were already a writer in). Preventing that requires that the message asserts its own sender, so that fixing the hash of the message via prev pointers fixes the sender, and that means we have to check it for consistency.

Followup thought: We thought about having signencrypt do this for us. The simplest possible solution, where we hash the signing key in with all the other associated data, doesn't actually work, because the weak signing key would still return true for the mismatching signature input. If we transport the public key alongside the plaintext (increasing the size of the final ciphertext, unlike the first idea), that could work, but the receiving client would have to remember to compare the keys. Another option would be to have the _encryption_ layer take the signing key as associated data, which clients can't forget to check. (NaCl's crypto_secretbox doesn't support associated data directly, but the standard way to do it is to hash the encryption key together with your AD, and use the result to encrypt the actual message, so that decryption breaks if the AD changes.)

Comment (mlsteele) [inline go/chat/boxer.go]: I hadn't thought of using the KID to look up the UID+DID. What you said makes sense. Here's another summary of what this does for v2.
The inputs are: `unsigned sender info`, `kid`, `signed sender info`. We check that the `kid` is valid for the `unsigned sender info` and then after opening the header we check that the `signed sender info` equals the `unsigned sender info`. Then we output `signed sender info` on `MessageClientHeaderVerified`.

I just noticed we don't check that the key is for the correct device. So I guess we should do that.

Comment (oconnor663) [inline go/chat/boxer.go]: Yep, that's the TODO on line 92.

File: go/chat/boxer_test.go [hunk 4]
```diff
@@ -340,38 +344,42 @@ func TestChatMessageInvalidHeaderSig(t *testing.T) {
...
+		require.True(t, called, "mangle must be called")
+
 		// need to give it a server header...
 		boxed.ServerHeader = &chat1.MessageServerHeader{
 			Ctime: gregor1.ToTime(time.Now()),
 		}
 
 		// put original signing fn back
-		boxer.sign = origSign
+		boxer.testingSignatureMangle = nil
```
Comment (oconnor663) [inline go/chat/boxer_test.go]: Why is this line needed?

File: go/chat/boxer.go [hunk 5]
```diff
@@ -402,8 +412,276 @@ func (b *Boxer) unboxV1(ctx context.Context, boxed chat1.MessageBoxed, encryptio
...
+		ctx, boxed.ClientHeader.Sender, boxed.HeaderVerificationKey, boxed.ServerHeader.Ctime)
+	if ierr != nil {
+		return nil, ierr
+	}
+	if !senderKeyFound {
+		return nil, NewPermanentUnboxingError(libkb.NoKeyError{Msg: "sender key not found"})
+	}
+	if !senderKeyValidAtCtime {
+		return nil, NewPermanentUnboxingError(libkb.NoKeyError{Msg: "key invalid for sender at message ctime"})
+	}
```
Comment (oconnor663) [inline go/chat/boxer.go]: Comment here about how this check is kind of a standin for really solving expired keys?

File: go/chat/boxer.go [hunk 6]
```diff
@@ -402,8 +412,276 @@ func (b *Boxer) unboxV1(ctx context.Context, boxed chat1.MessageBoxed, encryptio
...
+	return nil
+}
+
+func (b *Boxer) makeHeaderHash(ctx context.Context, headerSealed chat1.SignEncryptedData) (chat1.Hash, UnboxingError) {
+	buf := bytes.Buffer{}
+	err := binary.Write(&buf, binary.BigEndian, int32(headerSealed.V))
+	if err != nil {
+		return nil, NewPermanentUnboxingError(err)
+	}
+	_, err = buf.Write(headerSealed.N)
```
Comment (oconnor663) [inline go/chat/boxer.go]: yay

File: go/chat/boxer.go [hunk 7]
```diff
@@ -610,9 +889,60 @@ func (b *Boxer) boxV1(messagePlaintext chat1.MessagePlaintext, key *keybase1.Cry
...
+
+	bodyVersioned := chat1.NewBodyPlaintextWithV1(chat1.BodyPlaintextV1{
+		MessageBody: messagePlaintext.MessageBody,
+	})
+	bodyEncrypted, err := b.seal(bodyVersioned, encryptionKey)
+	if err != nil {
+		return nil, err
+	}
+
+	bodyHash := b.hashV1(bodyEncrypted.E)
```
Comment (oconnor663) [inline go/chat/boxer.go]: Since we're going through the trouble of computing the header hash with the nonce and version number thrown in, can we do the same here with the body hash?

**Comment by oconnor663** [comment]
We discussed changing the signature prefix to preclude any downgrading. That feels good to me, since all of our prefixes will still be hardcoded. We can add in proper associated data in the future (with another of these bumps) if we really need to.

LGTM with the comments above.

File: go/chat/boxer_test.go (hunk 4 above)
Comment (mlsteele) [inline go/chat/boxer_test.go]: Removed

File: go/chat/boxer.go (hunk 7 above)
Comment (mlsteele) [inline go/chat/boxer.go]: Added `makeBodyHash` used in mbv2. And a comment in avdl for what it does.

File: go/chat/boxer.go [hunk 8]
```diff
@@ -687,6 +690,23 @@ func (b *Boxer) makeHeaderHash(ctx context.Context, headerSealed chat1.SignEncry
 	return b.hashV1(buf.Bytes()), nil
 }
 
+func (b *Boxer) makeBodyHash(bodyCiphertext chat1.EncryptedData) (chat1.Hash, UnboxingError) {
+	buf := bytes.Buffer{}
+	err := binary.Write(&buf, binary.BigEndian, int32(bodyCiphertext.V))
+	if err != nil {
+		return nil, NewPermanentUnboxingError(err)
+	}
+	_, err = buf.Write(bodyCiphertext.N)
```
Comment (oconnor663) [inline go/chat/boxer.go]: I wonder if we should add an assertion in here that the nonce length has not changed? I know we're assuming it's a constant. (This would apply equally to the other one.)

File: go/protocol/chat1/extras.go [hunk 9]
```diff
@@ -208,10 +208,26 @@ func (o OutboxID) Eq(r OutboxID) bool {
 	return bytes.Equal(o, r)
 }
 
+// Compare pointers by value, not address.
+func (o *OutboxID) Eq2(r *OutboxID) bool {
```
Comment (oconnor663) [inline go/protocol/chat1/extras.go]: We talked about unifying `Eq` and `Eq2` by making them take a pointer receiver.

File: go/chat/boxer.go (hunk 8 above)
Comment (mlsteele) [inline go/chat/boxer.go]: Deferred to CORE-4723

File: go/protocol/chat1/extras.go (hunk 9 above)
Comment (mlsteele) [inline go/protocol/chat1/extras.go]: Done

**Comment by mlsteele** [comment]
Two new commits: the ones with (t). This drops `MessageBoxed.HeaderSealed` in favor of using `MessageBoxed.HeaderCiphertext` for both. `HeaderCiphertext` is now of type `HeaderSealed` which is basically a union of `HeaderEncrypted` and `HeaderSignEncrypted`. The client chooses how to interpret `HeaderCiphertext` based on the value of `MessageBoxed.Version`.

This change has two advantages:
- Saves some space. There won't be empty `HeaderCiphertext`s hanging out on V2 messages.
- The server side of this is less complicated. The server now doesn't need to decide which field to store and decode differently. It just passes around HeaderCiphertext same as before.

The reason for having 3 types instead of 1 is just so that the client can have more clear function signatures. For example, `func signEncrypt()` returns a `SignEncryptedData` even though in the end that becomes a `SealedData`.

Because I had already merged `MessageBoxed.HeaderSealed` (removed in this change), V1 messages from weeks around now will have an awkward empty one.
